File: Projects-Solutions/Numbers/Go/prime.go
/*Prime Factorization:
* Have the user enter a number and find all Prime Factors
* (if there are any) and display them.
 */
package main

import (
	"fmt"
	"math"
)

//primefactors : calculates and displays the prime factors of an integer
func primefactors(n int) {
	/* print number of twos that divide n*/
	for n%2 == 0 {
		fmt.Printf("%d ", 2)
		n /= 2
	}

	/*n must be odd hence i += 2
	 * all factor of n are less than sqrt of (n)*/
	sqrt_of_n := math.Sqrt(float64(n))
	for i := 3; i <= int(sqrt_of_n); i += 2 {
		for n%i == 0 {
			fmt.Printf("%d ", i)
			n /= i
		}
	}

	/* case where n is a prime number >  2*/
	if n > 2 {
		fmt.Printf("%d ", n)
	}

	fmt.Println()
}

func main() {
	var n int

	fmt.Println("Enter an integer to display it's prime factors")
	_, err := fmt.Scanf("%d ", &n)

	if err != nil {
		panic(err)
	}

	fmt.Printf("The prime factors of %d are: \n", n)
	primefactors(n)
}

//check out sieve of Eratosthenes (../../Classic-Algorithms/Go/sieve.go)

File: Projects-Solutions/Numbers/Java/Prime.java
/*Prime Factorization:
* Have the user enter a number and find all Prime Factors (if there are any) and display them.
*/
import java.util.Scanner;

public class Prime {
	public static void main(String[] args){
		Scanner input = new Scanner(System.in);
		int n;
		
		System.out.println("Enter number: ");
		n = input.nextInt();
		
		System.out.println("Prime factors of " + n + " are");
		primeFactors(n);

	}
	
	//primeFactors : finds prime factors of a number num
	public static void primeFactors(int num){
		int tmp = num;
		
		//if 2 is a prime factor of num
		while(tmp%2 == 0){
			System.out.print(2 + " ");
			tmp /= 2;
		}
		
		//find prime factors of n between 3 and sqrt(num)
		for(int i = 3; i <= Math.sqrt(num); i += 2){
			while(tmp%i == 0){
				System.out.print(i + " ");
				tmp /= i;
			}
		}
		
		//when num is a prime number
		if(num == tmp){
			System.out.println(num);
		}
		System.out.println();
	}
	
}
